# Лаборатоная работа №1

# Работа с библиотекой PANDAS



<div style = "font-size: 16px">
    <p><b>Цель работы: </b>познакомиться с использованием библиотеки pandas при решении практических задач, связанных с анализом данных<p>
    <br>
    <p>Проанализировать известный датасет student-mat. Датасет содержит сведения о учениках школы. Также он содержит информацию об успеваемости школьников, их здоровье, потреблении алкоголя, свободном времени и другие социальные факторы. Смысл задачи - выявление связей между этими факторами и успеваемостью студентов. Набор включает следующие атрибуты:</p>
    <ul>
    <li> <b>School</b> – школа ('GP' – Gabriel Pereira или 'MS' – Mousinho da Silveira);</li>
    <li> <b>Sex</b> – пол студента ('F' – female или 'M' – male);</li>
    <li> <b>Age</b> – возраст студента;</li>
    <li> <b>Adress</b> – домашний адрес студента ('U' – городской или 'R' – сельский);</li>
    <li> <b>Famsize</b> – размер семьи ('LE3' – меньше или равно 3 или 'GT3' – больше чем 3);</li>
    <li> <b>Pstatus</b> – статус совместного проживания родителей ('T' – совместное проживание или 'A' – раздельное проживание);</li>
    <li> <b>Medu</b> – образование матери (0 – нет, 1 – начальное (4 класс), 2-9 классы, 3 – среднее или 4 – высшее);</li>
    <li> <b>Fedu</b> – образование отца (0 – нет, 1 – начальное (4 класс), 2-9 классы, 3 – среднее или 4 – высшее);</li>
    <li> <b>Mjob</b> – работа матери (учитель, медицинское обслуживание, связанные с гражданскими услугами (административными или полицейскими), на дому или другое);</li>
    <li> <b>Fjob</b> – работа отца (учитель, медицинское обслуживание, связанные с гражданскими услугами (административными или полицейскими), на дому или другое);</li>
    <li> <b>Famrel</b> – качество семейных отношений (от 1 – очень плохо до 5 – отлично);</li>
    <li> <b>Freetime</b> – свободное время после школы (от 1 – очень мало до 5 – очень много);</li>
    <li> <b>Goout</b> – встречи с друзьями (от 1 – очень мало до 5 – очень много);</li>
    <li> <b>Dalc (Daily Alcohol Consumption)</b> – потребление алкоголя в рабочий день (от 1 – очень низкое до 5 – очень высокое);</li>
    <li> <b>Walc (Weekly Alcohol Consumption)</b> – еженедельное потребление алкоголя (от 1 – очень низкое до 5 – очень высокое);</li>
    <li> <b>Health</b> – текущее состояние здоровья (от 1 – очень плохое до 5 – очень хорошее);</li>
    <li> <b>Absences</b> – количество пропусков занятий в школе (от 0 до 93);</li>
    <li> <b>G1</b> – оценка за первый период (от 0 до 20);</li>
    <li> <b>G2</b> – оценка за второй период (от 0 до 20);</li>
    <li> <b>G3</b> – итоговая оценка (от 0 до 20, целевой результат).</li>
    </ul>
</div>


### 1. Импортируйте необходимые библиотеки

In [1]:
import pandas as pd
import numpy as np

### 2. Импортируйте набор данных из файла student-mat.csv и поместите его в объект DataFrame

In [2]:
df = pd.read_csv("student-mat.csv")

### 3. Выведите 16 первых строк

In [222]:
print(df.head(16))

   school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0      GP   F   18       U     GT3       A     4     4   at_home   teacher   
1      GP   F   17       U     GT3       T     1     1   at_home     other   
2      GP   F   15       U     LE3       T     1     1   at_home     other   
3      GP   F   15       U     GT3       T     4     2    health  services   
4      GP   F   16       U     GT3       T     3     3     other     other   
5      GP   M   16       U     LE3       T     4     3  services     other   
6      GP   M   16       U     LE3       T     2     2     other     other   
7      GP   F   17       U     GT3       A     4     4     other   teacher   
8      GP   M   15       U     LE3       A     3     2  services     other   
9      GP   M   15       U     GT3       T     3     4     other     other   
10     GP   F   15       U     GT3       T     4     4   teacher    health   
11     GP   F   15       U     GT3       T     2     1  services

### 4. Сколько строк и столбцов в наборе данных?

In [223]:
print("Количество строк:", len(df.axes[0]))
print("Количество столбцов:", len(df.axes[1]))

Количество строк: 395
Количество столбцов: 33


### 5. Получите сводную информацию по датафрейму

In [224]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      395 non-null    object
 1   sex         395 non-null    object
 2   age         395 non-null    int64 
 3   address     395 non-null    object
 4   famsize     395 non-null    object
 5   Pstatus     395 non-null    object
 6   Medu        395 non-null    int64 
 7   Fedu        395 non-null    int64 
 8   Mjob        395 non-null    object
 9   Fjob        395 non-null    object
 10  reason      395 non-null    object
 11  guardian    395 non-null    object
 12  traveltime  395 non-null    int64 
 13  studytime   395 non-null    int64 
 14  failures    395 non-null    int64 
 15  schoolsup   395 non-null    object
 16  famsup      395 non-null    object
 17  paid        395 non-null    object
 18  activities  395 non-null    object
 19  nursery     395 non-null    object
 20  higher    

### 6. Что является индексом этого набора данных?

In [225]:
df.index

RangeIndex(start=0, stop=395, step=1)

### 7. Создайте новый датафрейм, удалив все столбцы, связанные с семьей и происхождением студента
<br>
<p style = "font-size: 16px"> Уберем также столбцы относящиеся к другому заданию</p>

In [226]:
df_copy = df.copy()
drop_cols = ["famsize", "famrel", "Mjob", "Fjob", "Pstatus", "Medu", "Fedu", "romantic", "internet", "higher", "nursery", "activities", "paid", "famsup", "schoolsup", "failures", "studytime", "traveltime", "guardian", "reason"]
df_copy.drop(drop_cols, axis=1, inplace=True)
print(df_copy.head())

  school sex  age address  freetime  goout  Dalc  Walc  health  absences  G1  \
0     GP   F   18       U         3      4     1     1       3         6   5   
1     GP   F   17       U         3      3     1     1       3         4   5   
2     GP   F   15       U         3      2     2     3       3        10   7   
3     GP   F   15       U         2      2     1     1       5         2  15   
4     GP   F   16       U         3      2     1     2       5         4   6   

   G2  G3  
0   6   6  
1   5   6  
2   8  10  
3  14  15  
4  10  10  


### 8. По школам определите 15 студентов, имеющих наихудшую успеваемость, найдите для выделенной группы средние значения показателей health, Dalc и Walc

In [227]:
for col in df_copy["school"].unique():
    print("Список студентов из школы:", col)
    school_data = df_copy[df_copy['school'] == col].copy()
    school_data.sort_values(by="G3", inplace=True)
    worst_stud = school_data.head(15)
    print(worst_stud)
    print(f"Среднее значение показателя health: {worst_stud['health'].mean()}")
    print(f"Среднее значение показателя Dalc: {worst_stud['Dalc'].mean()}")
    print(f"Среднее значение показателя Walc: {worst_stud['Walc'].mean()}")
    print("\n")

Список студентов из школы: GP
    school sex  age address  freetime  goout  Dalc  Walc  health  absences  \
316     GP   F   18       U         3      3     1     2       1         0   
310     GP   F   19       U         2      4     2     2       3         0   
296     GP   F   19       U         3      4     2     3       2         0   
134     GP   M   15       R         3      3     1     1       5         0   
135     GP   F   15       U         3      3     1     1       5         0   
131     GP   F   15       U         3      3     1     2       4         0   
130     GP   F   15       R         2      2     2     2       5         0   
136     GP   M   17       R         4      5     2     4       5         0   
128     GP   M   18       R         3      3     1     2       4         0   
137     GP   F   16       U         3      2     1     1       5         0   
140     GP   M   15       U         2      2     1     1       3         0   
144     GP   M   17       U       

### 9. Определите 15 студентов, имеющих наилучшую успеваемость, найдите для выделенной группы средние значения показателей health, Dalc и Walc

In [228]:
for col in df_copy["school"].unique():
    print("Список студентов из школы:", col)
    school_data = df_copy[df_copy['school'] == col].copy()
    school_data.sort_values(by="G3", inplace=True, ascending=False)
    worst_stud = school_data.head(15)
    print(worst_stud)
    print(f"Среднее значение показателя health: {worst_stud['health'].mean()}")
    print(f"Среднее значение показателя Dalc: {worst_stud['Dalc'].mean()}")
    print(f"Среднее значение показателя Walc: {worst_stud['Walc'].mean()}")
    print("\n")

Список студентов из школы: GP
    school sex  age address  freetime  goout  Dalc  Walc  health  absences  \
47      GP   M   16       U         2      2     1     1       2         4   
113     GP   M   15       U         5      2     1     1       3        10   
110     GP   M   15       U         5      3     1     1       4         6   
286     GP   F   18       U         3      3     1     2       2         5   
8       GP   M   15       U         2      2     1     1       1         0   
303     GP   F   17       U         2      2     1     2       5         0   
104     GP   M   15       U         4      4     1     1       1         0   
306     GP   M   20       U         5      3     1     1       5         0   
36      GP   M   15       U         4      3     1     1       4         2   
42      GP   M   15       U         3      3     1     1       5         2   
91      GP   F   15       U         5      5     1     3       1         4   
260     GP   F   18       U       

### 10. Имеются ли в наборе лица, употребляющие спиртные напитки нелегально (возраст до 17 лет)?
<br>
<p style = "font-size: 16px"> Будем считать что при значении показателя Dalc и Walc = 1 студент не употребляет алкоголь
</p>

In [229]:
underage_students = df_copy[(df_copy['age'] < 17) & ((df_copy['Dalc'] > 1) | (df_copy['Walc'] > 1))].copy()
print("Количество студентов нелегально употребляющие спиртные напитки:", len(underage_students))

Количество студентов нелегально употребляющие спиртные напитки: 99


### 11. Определите число таких лиц в каждой школе. Кого больше, юношей или девушек?

In [230]:
for col in df_copy["school"].unique():
    print("Список студентов из школы:", col)
    underage_students_school = underage_students[underage_students['school'] == col].copy()
    print("Количество лиц в этой школе:", len(underage_students_school))
    m_cnt = len(underage_students_school[underage_students_school['sex'] == "M"])
    f_cnt = len(underage_students_school[underage_students_school['sex'] == "F"])
    print(f"Юношей: {m_cnt}\nДевушек: {f_cnt}")
    print("Юношей больше\n") if m_cnt > f_cnt else print("Девушек больше\n") if m_cnt < f_cnt else print("Одинаковое количество\n")

Список студентов из школы: GP
Количество лиц в этой школе: 99
Юношей: 52
Девушек: 47
Юношей больше

Список студентов из школы: MS
Количество лиц в этой школе: 0
Юношей: 0
Девушек: 0
Одинаковое количество



### 12. В какой школе наименьший процент студентов, злоупотребляющих алкоголем?
<br>
<p style = "font-size: 16px"> Будем считать что злоупотребление алкоголем считается при значении показателя Dalc или Walc > 3
</p>

In [231]:
for col in df_copy["school"].unique():
    print("Список студентов из школы:", col)
    total_students = df_copy[df_copy['school'] == col].copy()
    alco_students = total_students[(total_students['Dalc'] > 3) | (total_students['Walc'] > 3)]
    percent_alco = len(alco_students)/len(total_students) * 100
    print(f"Процент студентов, злоупотребляющих алкоголем {percent_alco}\n")

Список студентов из школы: GP
Процент студентов, злоупотребляющих алкоголем 20.34383954154728

Список студентов из школы: MS
Процент студентов, злоупотребляющих алкоголем 21.73913043478261



<p style = "font-size: 16px"> Исходя из результатов (21.739 > 20.343) можно увидеть что в школе MS процент студентов больше.

</p>

### 13. Связаны ли каким-то образом избыток свободного времени и частые встречи с друзьями со злоупотреблением алкоголем?
<br>
<p style = "font-size: 16px"> Добавим индикатор злоупотребления алкоголем (Alcohol_indicator) 1 для злоупотребляющих и 0 для остальных. И уберем остальные столбцы.
</p>

In [240]:
df_copy_alco = df_copy.copy()
df_copy_alco['Alcohol_indicator'] = ((df_copy_alco['Dalc'] > 3) | (df_copy_alco['Walc'] > 3)).astype(int)
drop_cols_alco = ['school', 'sex', 'age', 'address', 'Dalc', 'Walc', 'health', 'absences', 'G1', 'G2', 'G3']
df_copy_alco.drop(drop_cols_alco, axis=1, inplace=True)
freetime_analysis = df_copy_alco.groupby('freetime')['Alcohol_indicator'].mean()
print("Анализ по freetime:")
print(freetime_analysis)
goout_analysis = df_copy_alco.groupby('goout')['Alcohol_indicator'].mean()
print("\nАнализ по goout:")
print(goout_analysis)

Анализ по freetime:
freetime
1    0.105263
2    0.171875
3    0.165605
4    0.252174
5    0.325000
Name: Alcohol_indicator, dtype: float64

Анализ по goout:
goout
1    0.086957
2    0.087379
3    0.092308
4    0.337209
5    0.547170
Name: Alcohol_indicator, dtype: float64


<p style = "font-size: 16px"> Анализируя резульататы можно сказать что студенты у которых больше свободного время и частые встречи с друзьями больше склонны к злоупотреблению алкоголя.
</p>

### 14. Есть ли связь избытка свободного времени после школы и частых встреч с друзьями?
<br>
<p style = "font-size: 16px"> Оставим только столбцы freetime и goout.
</p>

In [251]:
df_copy_freetime = df_copy.copy()
drop_cols_freetime = ['school', 'sex', 'age', 'address', 'Dalc', 'Walc', 'health', 'absences', 'G1', 'G2', 'G3']
df_copy_freetime.drop(drop_cols_freetime, axis=1, inplace=True)
print(df_copy_freetime.groupby('freetime').mean())

             goout
freetime          
1         2.315789
2         2.718750
3         3.031847
4         3.408696
5         3.550000


<p style = "font-size: 16px"> Анализируя результаты можно сказать что чем больше свободного времени, тем выше среднее значение goout</p>

### 15. Получите сводку статистик для нечислового столбца Mjob


In [261]:
print(df['Mjob'].describe())

count       395
unique        5
top       other
freq        141
Name: Mjob, dtype: object
